player/Archive_Reader.cpp

```diff
@@ -14,45 +37,187 @@ static int CALLBACK call_rar( UINT msg, LPARAM UserData, LPARAM P1, LPARAM P2 )
+	if ( !(rar = RAROpenArchive( &data )) )
+		return erar_handle;
+	RARSetCallback( rar, call_rar, (LPARAM)&buf_ptr );
+	return nullptr;
 }
 
-gme_err_t Rar_Reader::read( void* p )
+gme_err_t Rar_Reader::next( void* bp, arc_entry_t* entry )
 {
-	bp = p;
-	if ( RARProcessFile( rar, -1, nullptr, nullptr ) != ERAR_SUCCESS )
-		return "Error processing RAR file";
+	// if prev entry was not a music emu file, buf_ptr returns to prev position
+	buf_ptr = bp;
+	int res;
+	if ( (res = RARReadHeader( rar, &head )) != ERAR_SUCCESS
+	|| (res = RARProcessFile( rar, RAR_TEST, nullptr, nullptr )) != ERAR_SUCCESS )
+		return erars[res - erar_offset];
+
+	entry->name = head.FileName;
+	entry->size = head.UnpSize;
 	return nullptr;
 }
 
+Rar_Reader::~Rar_Reader() {
+	RARCloseArchive( rar );
+}
+
 #endif // RARDLL
+
+
+#ifdef HAVE_LIBARCHIVE
+
+const char* zip_err_struct = "Failed to create archive struct";
+
+#if ARCHIVE_VERSION_NUMBER < 3000000
+#define archive_read_free archive_read_finish
+#define archive_read_support_filter_all archive_read_support_compression_all
+#endif
+
+#ifdef HAVE_ZLIB_H
+
+#include <zlib.h>
+
+static const uint16_t gz_signature = GME_2CHAR( 0x1f, 0x8b );
+
+static const int zerr_offset = -6;
+static gme_err_t const zerrs[] = {
+	"GZ: Bad version",       // Z_VERSION_ERROR (-6)
+	"GZ: Buffer too small",  // Z_BUF_ERROR    (-5)
+	"GZ: Out of memory",     // Z_MEM_ERROR    (-4)
+	"GZ: Bad Data",          // Z_DATA_ERROR   (-3)
+	"GZ: Stream error",      // Z_STREAM_ERROR (-2)
+};
+#endif // HAVE_ZLIB_H
+
+gme_err_t Zip_Reader::open_zip( const char* path ) {
+	if ( !(zip = archive_read_new()) )
+		return zip_err_struct;
+	if ( archive_read_support_filter_all( zip ) != ARCHIVE_OK
+	|| archive_read_support_format_zip( zip ) != ARCHIVE_OK
+	|| archive_read_open_filename( zip, path, 10240 ) != ARCHIVE_OK )
+		return archive_error_string( zip );
+	return nullptr;
+}
+
+gme_err_t Zip_Reader::open( const char* path )
+{
+	gme_err_t err;
+	if ( (err = open_zip( path )) )
+		return err;
+
+	// determine space needed for the unpacked size and file count.
+	int res;
+	while ( (res = archive_read_next_header( zip, &head )) == ARCHIVE_OK )
+	{
+#ifdef HAVE_ZLIB_H
+		char h[3];
+		archive_read_data( zip, &h, 3 );
+		if ( GME_2CHAR( h[0], h[1] ) == gz_signature && h[2] == 8 )
+		{
+			// gzip puts its uncompressed file size in the footer
+			gme_vector<uint8_t> buf;
+			if ( (err = buf.resize( archive_entry_size( head ) - 3 )) )
+				return err;
+			archive_read_data( zip, buf.begin(), buf.size() );
+			const uint8_t* b = buf.end() - 4;
+			size_ += GME_4CHAR(b[3], b[2], b[1], b[0]);
+		}
+		else
+#endif // HAVE_ZLIB_H
+		{
+			size_ += archive_entry_size( head );
+		}
+		count_ += 1;
+	}
+	if ( res != ARCHIVE_EOF || archive_read_free( zip ) != ARCHIVE_OK )
+		return archive_error_string( zip );
+	return (err = open_zip( path )) ? err : nullptr;
+}
+
+gme_err_t Zip_Reader::next( void* buf_ptr, arc_entry_t* entry )
+{
+	ssize_t res;
```

Comment:
`ssize_t` may not be universally available, e.g. MSVC. (Although the CI MSVC runs seem to have succeeded...)

Reply:
what about `ptrdiff_t`?

Reply:
ptrdiff_t is available to many, yes.